### PR TITLE
feat: improved project initialization settings

### LINKED ISSUE #557: [Meta] Synthetics Agent enhancements for Private Node
To support private node, and enhance the current experience of the project monitors, updates are needed to the synthetics `init` command, `monitor.use` DSL, and `push` command line arguments.

Enhancements to `init`
--
- [ ] The new init command will update the existing questionnaire to the following options. Exact content/phrasing TBD.

* Are you using elastic cloud? Y/N
     Y: What is your cloud.id (You can find this....) ? 
     N: What is the url of your Kibana instance?
* What api key would you like to use for auth? (Include link to Kibana to set up API Key)
* Choosing a default location
     * IF locations are available, user will be able to select froma list of locations that includes both publc (if in cloud), and private locations together, with each location clearly indicating if it's public or private. 
     * IF locations are not available (for example, an on-prem cluster that has not yet set up private locations): A link to docs for how to set up your first private location is displayed
 * Choose default interval
 * Choose project id
 * Choose space
 
DSL experience with `monitor.use`
--
 - [x] Add new key for private locations `privateLocations`
 
`push` command override
--
- [x] Add new --private-locations option to the push command

Locations discovery
--
- [ ] Provide additional discovery for new locations via new synthetics discovery command: https://github.com/elastic/synthetics/issues/551

### PR DESCRIPTION
**Pull request by vigneshshanmugam**
+ fix #557 
+ fix  https://github.com/elastic/synthetics-dev/issues/130
+ Improves the synthetics project initialization experience with new sets of fields that helps streamline the experience for new users to onboard and start creating and running monitors seamlessly in Synthetics service and also on private nodes. 
+ The way it works is to ask set of questions to the users around how the users are running the Elastic stack and extracts the Kibana URL if the user is on cloud and requires API Key 
  - To fetch all location metadata from the Kibana
  - Creates a Project settings under `.synthetics/project.json`
  - Uses the scaffolded project skeleton along with `project.json` and `synthetics.config.ts` as source of truth
  - Errors on pushes with incorrect project settings
  - Allow users to manually override the values with CLI flags or via DSL.
+ If the user has already scaffolded the project using prior version of Synthetics, Installing the latest version would ask the users to `Initialize` the project again to establish the default settings. Users can follow along and manually override specific files if they want to or skip steps to onboard faster. 
+ Pushing the monitors via different project name prompts the user explicitly and follows the user input. 

### Testing.
1. Build the local project using - `npm run build`
2. Setup a brand new Synthetics project - `node dist/cli.js init [optional-dir]`
3. Follow along 
     - Elastic Cloud vs On Prem customers
     - Public locations & Private Locations
     - Kibnana Auth
4. Pushing monitors with default settings - only required arg is `auth`. 
5. Try different scenarios like pushing under different project name - `SYNTHETICS_API_KEY="" npm run push -- --project different-name`
6. Delete schedule and locations and try again to experience different project settings related errors. 
7. Go to any existing scaffolded synthetics project using prior versions of the agent and try pushing to experience different errors. For this to work properly. You would need to link to the current synthetics version. You can do so by changing `package.json` to `@elastic/synthetics: file:/path/to/locally/built/synthetics`

### CODE REVIEW

File: src/cli.ts [hunk 1]
```diff
@@ -229,4 +233,23 @@ program
...
+  .description('List all locations to run the synthetics monitors')
+  .option('--url <url>', 'kibana URL to upload the monitors')
+  .requiredOption(
+    '--auth <auth>',
+    'API key used for Kibana authentication(https://www.elastic.co/guide/en/kibana/master/api-keys.html).'
+  )
+  .action(async (cmdOpts: LocationCmdOptions) => {
+    try {
+      console.log(await getLocations(cmdOpts));
+      // TODO: Display list of locations
```
Comment (vigneshshanmugam) [inline src/cli.ts:248]: This will be covered as part of - https://github.com/elastic/synthetics/issues/551

**Review by lucasfcosta — COMMENTED** [review]
I've just followed the instructions specified both for a cloud deployment and for a local deployment.

I also tried a variation of pushing to custom locations and existing managed locations, and multiple different projects, that worked fine.

As expected, changing the project resulted in a "push aborted".

The only problem I found was that when I interactively select a private location it's not actually shown as selected after. [See this video to understand](https://user-images.githubusercontent.com/6868147/182426709-45e4e95c-d4b4-4121-8b24-c6d61652825c.mov).


Just before I re-review, I just wanna understand better what you mean with this testing step:

```
Go to any existing scaffolded project and try pushing to experience different errors. For this to work properly. You would need to link to the current synthetics version. You can do so by changing package.json to @elastic/synthetics: file:/path/to/locally/built/synthetics
```

What exactly does it mean to go into any scaffolded project and try pushing? Shouldn't the push work within the scaffold already?

If with this paragraph you mean it shouldn't work without pointing to the locally built synthetics then it's okay.

**Comment by vigneshshanmugam** [comment]
> The only problem I found was that when I interactively select a private location it's not actually shown as selected after. [See this video to understand](https://user-images.githubusercontent.com/6868147/182426709-45e4e95c-d4b4-4121-8b24-c6d61652825c.mov).

Did you click `Enter` ? I think since its a multiselect, need to do it via `Space`. Hmmm this feels like we need to tweak the settings to make Enter select a single item. Can you try it again? 

>What exactly does it mean to go into any scaffolded project and try pushing? Shouldn't the push work within the scaffold already?

I was trying to say, If you already have an existing Synthetics project scaffolded from priort synthetics versions. Running `init` with this PR should overwrite all default settings and make your project ready for working with private locations.

**Comment by vigneshshanmugam** [comment]
> Did you press Enter ? I think since its a multiselect, need to do it via Space. Hmmm this feels like we need to tweak the settings to make Enter select a single item. Can you try it again?

Added hints now which should inform users on how to select the locations. Also added validation to pick at-least one default location.

**Review by dominiqueclarke — CHANGES_REQUESTED** [review]
Hi there. I noticed a few things.

Private locations configuration
---
Screen Shot 2022-08-02 at 2 47 06 PM

Default private locations are stored in `synthetics.config.ts` with the `(private)` post fix. This will not work as Kibana will not be able to find the associated location. When pushing, my monitors indeed were not assigned to the private location. 

Trailing slashes
--
Screen Shot 2022-08-02 at 2 42 41 PM
We should probably remove trailing slashes when formatting KIbana url

Cloud flow
--
We discussed using cloud id vs Kibana url to find the Kibana instance. Is that still the plan, or did we want to go with Kibana url as an MVP first?
Screen Shot 2022-08-02 at 2 39 49 PM

**Comment by vigneshshanmugam** [comment]
@dominiqueclarke Fixed all the issues mentioned in the review. Please have another look.

**Review by lucasfcosta — CHANGES_REQUESTED** [review]
I've just tested this and found one more problem.

Now, it seems I always need to specify a `--url` value although that's already defined in the `.synthetics` file of my project.

```
{
  "url": "https://my-example.us-west2.gcp.elastic-cloud.com:443",
  "space": "default",
  "project": "examplelocal2"
}
```

Here's what I see:

```
➜ SYNTHETICS_API_KEY="<REDACTED>" npm run push -- --locations "japan" --project test

> examplelocal3@1.0.0 push
> npx @elastic/synthetics push --auth $SYNTHETICS_API_KEY "--locations" "japan" "--project" "test"

error: required option '--url <url>' not specified
```

Before the most recent changes I could do this just fine. If I manually use the URL as the URL param, the push works just fine 🤔 

I also double checked the new defaults and the changes to the CLI UI and they LGTM, it's just the URL thing now.

**Review by lucasfcosta — APPROVED** [review]

**Comment by lucasfcosta** [comment]
After re-testing with a new build it worked! My bad for not pointing the NPM script to the right place.

**Review by dominiqueclarke — APPROVED** [review]
LGTM. Concerns addressed.

**Pull request merged**